d2l-image.js

```diff
@@ -1,81 +1,114 @@
-import '@polymer/polymer/polymer-legacy.js';
-import '@polymer/iron-ajax/iron-ajax.js';
-import { Polymer } from '@polymer/polymer/lib/legacy/polymer-fn.js';
-const $_documentContainer = document.createElement('template');
-
-$_documentContainer.innerHTML = `<dom-module id="d2l-image">
-	<template strip-whitespace="">
-		<style>
+import 'd2l-fetch/d2l-fetch.js';
+import { css, html, LitElement, nothing } from 'lit';
+import { ifDefined } from 'lit/directives/if-defined.js';
+
+class D2LImage extends LitElement {
+	static get properties() {
+		return {
+			/**
+			 * Alternate text describing the image
+			 * @type {string}
+			 */
+			alternateText: { type: String, attribute: 'alternate-text' },
+			/**
+			 * URL of the image
+			 * @type {string}
+			 */
+			imageUrl: { type: String, attribute: 'image-url' },
+			/**
+			 * token used to authenticate the image retrieval
+			 * @type {string}
+			 */
+			token: { type: String },
+			_imageUrl: { state: true }
+		};
+	}
+
+	static get styles() {
+		return css`
 			:host {
 				display: block;
 			}
 
 			img {
-				width:100%;
-				height: 100%;
 				border-radius: var(--d2l-image-border-radius);
+				height: 100%;
 				object-fit: var(--d2l-image-object-fit);
+				width: 100%;
 			}
-		</style>
-
-		<iron-ajax id="imageRequest" url="[[imageUrl]]" headers="[[_headers]]" on-iron-ajax-response="_onImageResponse" on-iron-ajax-error="_onImageResponse" handle-as="blob">
-		</iron-ajax>
-
-		<img id="image" alt$="[[alternateText]]">
-	</template>
-
-</dom-module>`;
-
-document.head.appendChild($_documentContainer.content);
-
-Polymer({
-	is: 'd2l-image',
-
-	properties: {
-		alternateText: {
-			type: String,
-			reflectToAttribute: true,
-		},
-		imageUrl: {
-			type: String,
-			value: null
-		},
-		token: String,
-		_headers: String
-	},
-
-	observers: [
-		'_onImageUrlChange( imageUrl, token )'
-	],
-
-	_onImageUrlChange: function(imageUrl, token) {
-		if (imageUrl && token) {
-			this._headers = {
-				Authorization: 'Bearer ' + token
-			};
-			this.$.imageRequest.generateRequest();
-		} else if (imageUrl) {
-			this.$.image.src = imageUrl;
+		`;
+	}
+
+	constructor() {
+		super();
+		this.imageUrl = undefined;
+		this.alternateText = undefined;
+	}
+
+	render() {
+		if (!this._imageUrl) {
+			return nothing;
 		}
-	},
 
-	_onImageResponse: function(response) {
-		if (response.detail.status === 200) {
-			if (this.$.image.src) {
-				URL.revokeObjectURL(this.$.image.src);
+		return html`
+			<img alt="${this.alternateText}" src="${ifDefined(this._imageUrl)}">
+		`;
+	}
+
+	updated(changedProperties) {
+		if (changedProperties.has('imageUrl') || changedProperties.has('token')) {
+			this._loadImage();
```

Comment:
I've seen Lit behave a little weird in the past, where the order that the consumer specifies the attributes defines the order that `updated` gets calls. And in your case, if `updated` is first called with `imageUrl` before it's called with `token`, then `_loadImage` will attempt to load it without using the token.

So if a consumer did this, I think it might break: `<d2l-image image-url="foo" token="bar">`.

I'm not really sure how to get around that, since this is designed such that the token is optional. I think the old Polymer version likely suffered from something similar.

Reply:
Do you know if the updates queue? So like `updated` could be called without `token`, but then called again with `token` afterwards?

Reply:
So they're _supposed_ to queue... like if you go `elem.foo = 'bar'; elem.blah = 1;` that should just trigger one update. But I've definitely seen when a component is booting up and parsing the attributes and applying them to the properties that it can get called multiple times.